Create the initial version of optional_query_cache

SteveMarshall opened the pull request:
This is a drop-in replacement for ActiveRecord::QueryCache that allows query-caching to be disabled on specific routes.

This is particularly useful for routes used to report if the app can connect to the database or not (which ActiveRecord::QueryCache would normally block by way of a connection error).

SteveMarshall committed:
Apply the correct license to the code

SteveMarshall committed:
Add a basic README

There are a few assumptions here, based on my limited knowledge of
Rails middleware and gems. I'll probably need to change the info on
getting started once I've actually got the thing working.

SteveMarshall committed:
Add the basic gem framework

SteveMarshall committed:
Allow OptionalQueryCache to replace QueryCache

With no parameters or special configuration, we should behave exactly
like ActiveRecord::QueryCache.

SteveMarshall committed:
Allow query caching to be disabled for routes

This can be used by setting `disable_query_cache: true` on a route:

    # config/routes.rb
    get 'healthcheck', to: 'healthcheck#index', disable_query_cache: true

Or using a scope to disable query caching for multiple routes:

    scope disable_query_cache: true do
      # routes
    end

SteveMarshall committed:
Document disabling query cache for multiple routes

SteveMarshall committed:
Permit caching for unrecognised routes

ActionDispatch::Routing::RouteSet.recognize_path has a number of cases
where it raises ActionController::RoutingErrors. We should handle these
gracefully, ensuring that they don't disable query caching
unnecessarily.

SteveMarshall commented on lib/optional_query_cache/version.rb:
```diff
@@ -0,0 +1,11 @@
+require 'active_record'
```
This breaks `bundle install` on CI environments, but is required for `OptionalQueryCache` to not have a  superclass mismatch. Hmm.

tatyree replied:
What is the stack trace when it breaks?  Also, I'm not sure `version.rb` is the right place for this to live...

tatyree replied:
Sorry, I just re-read your comment...how very strange.  Can it be reproduced locally?

SteveMarshall replied:
You're right: it shouldn't live here, but making `OptionalQueryCache` a module with a `Middleware` class inside it seems… clunky to me?

Basically, what's happening is the gemspec tries to require this at `bundle install` time (when running on CI) and can't because it hasn't run `bundle install` yet… Catch 22 :boom:

tatyree replied:
Yeah, I've got it checked out now and am pondering...Tricky.

mloughran replied:
Rather than requiring active record, you can just declare the types:

```
module ActiveRecord; class QueryCache; end; end
```

threedaymonk replied:
@mloughran's solution is the way to go if you're set on the current hierarchy, but I think the problem is the fact that your namespace is a subclass. The alternative would be to create a separate top-level module as a namespace for everything, e.g.:

``` ruby
module OptionalQueryCache
  class Version
    # ...
  end
end

module OptionalQueryCache
  class QueryCache < ActiveRecord::QueryCache
    # ...
  end
end
```

SteveMarshall replied:
Yeah, @mloughran suggested that structure on Friday; I'll probably do exactly that this morning.

mattpep commented on LICENCE:
```diff
@@ -0,0 +1,19 @@
+Copyright (C) 2015 Crown copyright (Ministry of Justice)
```
Should probably be 2016.

SteveMarshall replied:
Whoops! Fixed.

SteveMarshall committed:
Properly namespace the main object

Having the root namespace as the main class means we have to apply its
superclass throughout the entire namespace. This fixes that by moving
the main object to `OptionalQueryCache::QueryCache`.

This caused building/running the repository on CI services to fail
because they couldn't `bundle install` without parsing the `gemspec`,
which relied on parsing `OptionalQueryCache::Version`, which in turn
depended on `active_record`… which hadn't been installed because
`bundle install` hadn't yet been run.

tatyree commented on spec/optional_query_cache_spec.rb:
```diff
@@ -0,0 +1,65 @@
+    env = double
+    allow(env).to receive(:[]).with('REQUEST_PATH').and_return('')
+    env
+  end
```
This would be neater as a tap:

```
let :env do
  double.tap |env|
    allow(env).to receive(:[]).with('REQUEST_PATH').and_return('')
  end
end
```

SteveMarshall replied:
:+1:

tatyree commented:
LGTM once you squash the fixups.  @threedaymonk?

tatyree commented:
and @mloughran?

aldavidson commented on spec/optional_query_cache_spec.rb:
```diff
@@ -0,0 +1,65 @@
+  end
+
+  it 'permits query caching by default' do
+    expect(ActiveRecord::Base).to receive(:connection).and_return(
```
two expectations in the same example is usually a bad sign - is this really an expectation, or a stub?

SteveMarshall replied:
A good point; the first of these is the expectation, the second (`[action_dispatch.routes]`) is a stub.

SteveMarshall replied:
Fixed!

aldavidson replied:
also, in this case, it looks like nothing else needs to be done with the return value, so the double is redundant?

SteveMarshall replied:
Nothing's done by this gem, no, but the double exists to stop `ActiveRecord::QueryCache` blowing up when it gets `call`ed as the super of `OptionalQueryCache::QueryCache`.

SteveMarshall replied:
That said, it strikes me that that's a good reason to put the doubles for `ActiveRecord::Base.connection` into `let()`. I'll do that.

SteveMarshall committed:
fixup! Allow query caching to be disabled for routes

SteveMarshall committed:
fixup! Permit caching for unrecognised routes

aldavidson commented on spec/optional_query_cache_spec.rb:
```diff
@@ -0,0 +1,65 @@
+    subject.call(env)
+  end
+
+  it 'permits query caching for unrecognised routes' do
```
when you've got two or more very similar examples which test the same thing with different parameters, its often clearer to structure it with context blocks, like:

``` ruby
  context "when the route is unrecognised" do
    # set up context that emulates the route being unrecognized, e.g.
    before do
      allow(env).to receive_message_chain(:[],:recognize_path).and_raise(
        ActionController::RoutingError.new('Unrecognised route')
      )
    end
    it "permits query caching" do
      expect(ActiveRecord::Base).to receive(:connection).and_return(double('mock connection'))
    end
  end

  context "when the route is specified" do
    before do
      allow(env).to receive(:[]).with('action_dispatch.routes').and_return(
        double(
          'routes',
          recognize_path: {
            disable_query_cache: true
          }
        )
      )
    end

    it "disables query caching" do
      expect(ActiveRecord::Base).not_to receive(:connection)
    end
  end
```

SteveMarshall replied:
A good point; fixed!

aldavidson commented on spec/optional_query_cache_spec.rb:
```diff
@@ -0,0 +1,65 @@
+
+  it 'disables query caching for specified routes' do
+    expect(ActiveRecord::Base).not_to receive(:connection)
+    expect(env).to receive(:[]).with('action_dispatch.routes').and_return(
```
pretty sure this is a stub, not an expectation, am i missing something?

SteveMarshall committed:
fixup! Allow OptionalQueryCache to replace QueryCache

SteveMarshall committed:
fixup! Allow query caching to be disabled for routes

SteveMarshall committed:
fixup! Allow query caching to be disabled for routes

SteveMarshall committed:
fixup! Permit caching for unrecognised routes

SteveMarshall committed:
Explicitly use a common connection mock

Both places where we were using a double for the connection should be
using the same one, and we should be more explicit about the fact that
it's standing in for a real connection (which should only be used by
the real QueryCache).

SteveMarshall committed:
fixup! Explicitly use a common connection mock

aldavidson commented:
LGTM  - I think the spec is much clearer now
